### PR TITLE
Improve diagnostics for unreadable artifacts

### LINKED ISSUE #27188: ArchiveArtifacts fails if one of the file being archived does not have the read attribute. Unix only.
### Jenkins and plugins versions report

Jenkins crashes when ArchiveArtifacts is used and one of the files being archived is missing the read attribute. Expected is a message explaining that a file is skipped rather than a crash. This has taken more than a man month to find out the issue at the Rocket Software with the uniface build pipeline. 

This crash only happens on Unix systems since there files can lack this read-attribute.
A fix which makes it possible to see which files are skipped with reasoning would be very much appreciated.

Thanks in advance
Jasper de Keijzer
Principal Software engineer at Rocket Software.

### What Operating System are you using (both controller, and any agents involved in the problem)?

We use all OS's supported by Jenkins. Uniface is platform independent.

### Reproduction steps

Use ArchiveArtifacts on a folder containing files which no read attribute set.
This will result in a crash.

### Expected Results

No crash and a clear reason why a file is not part of the archive. Or at least the option to get this info so it can be logged.

### Actual Results

Crash as explained.

### Anything else?

_No response_

### Are you interested in contributing a fix?

I have no time.

### PR DESCRIPTION
**Pull request by Hardik180704**
Fixes #27188

`ArtifactArchiver` previously reduced an `AccessDeniedException` to its raw exception string. Although this identified the denied path, it did not explain which archive operation failed, identify the workspace, or tell administrators which permissions to check.

For non-empty unreadable files, `TarArchiver` previously wrote the tar header before opening the file. When opening failed, closing the incomplete entry replaced the original `AccessDeniedException` with tar and EOF errors. This change opens regular files before writing their tar headers, preserving the actionable access-denied failure without leaving an incomplete entry.

`ArtifactArchiver` now reports the denied path and workspace and directs administrators to verify the controller and agent file permissions. It deliberately preserves the existing fail-safe behavior: the build still fails rather than reporting an incomplete artifact set as successfully archived. The detailed exception remains available in the `hudson.tasks.ArtifactArchiver` logger at `FINE` level, while the build log remains concise.

The existing Unix-only regression test now covers the actionable message and is associated with both #22902 and #27188.

### Testing done

Ran the focused regression test through the complete Maven reactor, including the Jenkins WAR and a Unix agent:

```text
mvn -ntp -pl test -am -Dtest=hudson.tasks.ArtifactArchiverTest#archiveNotReadable -Dskip.frontend=true -Dyarn.lint.skip=true package
```

Result:

```text
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0
BUILD SUCCESS
```

Also ran the dedicated tar archiver tests:

```text
mvn -ntp -pl core -am -Dtest=hudson.util.io.TarArchiverTest -Dskip.frontend=true -Dyarn.lint.skip=true package
Tests run: 4, Failures: 0, Errors: 0, Skipped: 1
BUILD SUCCESS
```

The test verifies that an unreadable matched artifact:

- keeps the build result as `FAILURE`;
- logs the denied path and workspace with permission guidance;
- avoids the incomplete tar-entry and EOF stack traces reproduced with a non-empty file;
- does not print a long stack trace.

### Screenshots (UI changes only)

Not applicable; this changes build-log diagnostics only.

#### Before

Not applicable.

#### After

Not applicable.

### Proposed changelog entries

- Improve artifact archiving diagnostics when file access is denied.

### Proposed changelog category

/label bug

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] The issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see examples in the Jenkins changelog).
- [x] There is automated testing or an explanation as to why this change has no tests.
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate. No public API was added.
- [x] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable. No deprecations were added.
- [x] UI changes do not introduce CSP regressions. No UI or JavaScript changes were made.
- [x] Dependency updates include links to external changelogs and full differentials where possible. No dependencies were updated.
- [x] New APIs and extension points link to at least one consumer. No APIs or extension points were added.

### Desired reviewers

N/A

Before the changes are marked as `ready-for-merge`:

### Maintainer checklist

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps for users, the `upgrade-guide-needed` label is set and the **Proposed upgrade guidelines** section is complete.
- [ ] If it would make sense to backport the change to LTS, the issue or pull request is labeled `lts-candidate` for consideration.

### CODE REVIEW

**Review by MarkEWaite — CHANGES_REQUESTED** [review]
I saw no difference in the output with this change compared to the current code on the master branch.  My test environment is included in  [gh-core-27188.zip](https://github.com/user-attachments/files/30789762/gh-core-27188.zip)

The steps that I took while trying to duplicate the issue:

1. Download and unzip [gh-core-27188.zip](https://github.com/user-attachments/files/30789762/gh-core-27188.zip) on my Linux computer
2. Change to the gh-core-27188 directory and run `bash ./README`
3. Open a web browser on port 8080 of the Linux computer 
4. Login with my Linux username ('mwaite') and the password 'p-' plus my username ('p-mwaite')
5. Review the results of the job 'archiveArtifacts-fails-on-unreadable-files'.  Confirm that it reports a stack trace:
```
Started by user SYSTEM
[Pipeline] Start of Pipeline
[Pipeline] node
Running on rhel-8-a.markwaite.net-mwaite in /home/mwaite/rhel-8-a.markwaite.net-mwaite/workspace/archiveArtifacts-fails-on-unreadable-files
[Pipeline] {
[Pipeline] stage
[Pipeline] { (Archive an unreadable file)
[Pipeline] sh
+ date
[Pipeline] sh
+ date
+ chmod 000 date-file-2.txt
[Pipeline] archiveArtifacts
Archiving artifacts
[Pipeline] }
[Pipeline] // stage
[Pipeline] }
[Pipeline] // node
[Pipeline] End of Pipeline
Also:   	Also:   hudson.remoting.ProxyException: java.io.IOException: This archives contains unclosed entries.
			at Jenkins v2.568.2//org.apache.tools.tar.TarOutputStream.finish(TarOutputStream.java:235)
			at Jenkins v2.568.2//org.apache.tools.tar.TarOutputStream.close(TarOutputStream.java:252)
			at Jenkins v2.568.2//hudson.util.io.TarArchiver.close(TarArchiver.java:128)
			at Jenkins v2.568.2//hudson.FilePath.writeToTar(FilePath.java:3053)
	Also:   hudson.remoting.ProxyException: hudson.remoting.Channel$CallSiteStackTrace: Remote call to rhel-8-a.markwaite.net-mwaite
			at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1916)
			at hudson.remoting.UserRequest$ExceptionResponse.retrieve(UserRequest.java:384)
			at hudson.remoting.Channel$2.adapt(Channel.java:1144)
			at hudson.remoting.Channel$2.adapt(Channel.java:1140)
			at hudson.remoting.FutureAdapter.get(FutureAdapter.java:66)
			at hudson.FilePath.copyRecursiveTo(FilePath.java:2850)
			at jenkins.model.StandardArtifactManager.archive(StandardArtifactManager.java:73)
			at hudson.tasks.ArtifactArchiver.perform(ArtifactArchiver.java:258)
			at PluginClassLoader for workflow-basic-steps//org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:101)
			at PluginClassLoader for workflow-basic-steps//org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:71)
			at PluginClassLoader for workflow-step-api//org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:49)
			at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:545)
			at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:328)
			at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1090)
			at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:614)
			at java.base/java.lang.Thread.run(Thread.java:1474)
hudson.remoting.ProxyException: java.io.IOException: entry 'date-file-2.txt' closed at '0' before the '29' bytes specified in the header were written
		at Jenkins v2.568.2//org.apache.tools.tar.TarOutputStream.closeEntry(TarOutputStream.java:362)
		at Jenkins v2.568.2//hudson.util.io.TarArchiver.visit(TarArchiver.java:121)
		at Jenkins v2.568.2//hudson.util.DirScanner.scanSingle(DirScanner.java:52)
		at Jenkins v2.568.2//hudson.FilePath$ExplicitlySpecifiedDirScanner.scan(FilePath.java:3734)
		at Jenkins v2.568.2//hudson.FilePath.writeToTar(FilePath.java:3054)
		at Jenkins v2.568.2//hudson.FilePath$CopyRecursiveRemoteToLocal.invoke(FilePath.java:3019)
		at Jenkins v2.568.2//hudson.FilePath$CopyRecursiveRemoteToLocal.invoke(FilePath.java:3002)
		at Jenkins v2.568.2//hudson.FilePath$FileCallableWrapper.call(FilePath.java:3655)
		at hudson.remoting.UserRequest.perform(UserRequest.java:225)
		at hudson.remoting.UserRequest.perform(UserRequest.java:50)
		at hudson.remoting.Request$2.run(Request.java:391)
		at hudson.remoting.InterceptingExecutorService.lambda$wrap$0(InterceptingExecutorService.java:81)
		at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
		at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
		at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
		at java.base/java.lang.Thread.run(Thread.java:1583)
	Caused: hudson.remoting.ProxyException: java.util.concurrent.ExecutionException: java.io.IOException: entry 'date-file-2.txt' closed at '0' before the '29' bytes specified in the header were written
		at hudson.remoting.Channel$2.adapt(Channel.java:1146)
		at hudson.remoting.Channel$2.adapt(Channel.java:1140)
		at hudson.remoting.FutureAdapter.get(FutureAdapter.java:66)
		at hudson.FilePath.copyRecursiveTo(FilePath.java:2850)
Also:   hudson.remoting.ProxyException: org.jenkinsci.plugins.workflow.actions.ErrorAction$ErrorId: 7ff33e96-e9ff-45ce-9f71-0c717645a5c7
hudson.remoting.ProxyException: java.io.EOFException
	at java.base/java.util.zip.GZIPInputStream.readUByte(GZIPInputStream.java:332)
	at java.base/java.util.zip.GZIPInputStream.readUShort(GZIPInputStream.java:322)
	at java.base/java.util.zip.GZIPInputStream.readHeader(GZIPInputStream.java:227)
	at java.base/java.util.zip.GZIPInputStream.<init>(GZIPInputStream.java:86)
	at java.base/java.util.zip.GZIPInputStream.<init>(GZIPInputStream.java:118)
	at hudson.FilePath$TarCompression$2.extract(FilePath.java:890)
	at hudson.FilePath.copyRecursiveTo(FilePath.java:2847)
	at jenkins.model.StandardArtifactManager.archive(StandardArtifactManager.java:73)
	at hudson.tasks.ArtifactArchiver.perform(ArtifactArchiver.java:258)
	at PluginClassLoader for workflow-basic-steps//org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:101)
	at PluginClassLoader for workflow-basic-steps//org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:71)
	at PluginClassLoader for workflow-step-api//org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:49)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:545)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:328)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1090)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:614)
	at java.base/java.lang.Thread.run(Thread.java:1474)
Finished: FAILURE
```

The job definition that created that output is:

```
pipeline {
    agent any
    stages {
        stage('Archive an unreadable file') {
            steps {
                sh 'date >> date-file.txt'
                sh 'date >> date-file-${BUILD_ID}.txt && chmod 000 date-file-${BUILD_ID}.txt'
                archiveArtifacts artifacts: 'date-file*.txt'
            }
        }
    }
}
```

**Comment by Hardik180704** [comment]
@MarkEWaite 
Thanks for the detailed reproduction, Mark. You were right—the existing test used a zero-byte unreadable file, so it did not expose the failure shown in your environment.
I updated the test to use a non-empty unreadable file and reproduced the same incomplete tar-entry and EOF stack traces. The root cause was that TarArchiver wrote the tar header before opening the file, allowing the original AccessDeniedException to be masked when the incomplete entry was closed.
Commit be054f74b9 now opens regular files before writing their tar headers, preserving the original access-denied failure. I verified it with the strengthened Unix regression test and the dedicated TarArchiverTest suite; both pass successfully.
Thanks again for providing the test environment and identifying the gap in our original reproduction.

**Review by MarkEWaite — CHANGES_REQUESTED** [review]
The modified lines in TarArchiver are not reached by any test in the ArtifactArchiverTest.  When I use a debugger to run the automated tests that are changed in this pull request, with  breakpoints on each line that was changed in TarArchiver, my breakpoints are never encountered by the automated test.

Please be sure that your automated tests execute the modified lines in the production code.

**Comment by Hardik180704** [comment]
@MarkEWaite 
Thanks for pointing this out. I added a direct in-process regression test, `TarArchiverTest.unreadableFileDoesNotLeaveIncompleteEntry`, which invokes `TarArchiver.visit` with a non-empty unreadable file and verifies that the original `AccessDeniedException` is preserved without leaving an incomplete tar entry.

I also kept the existing `ArtifactArchiverTest` for end-to-end coverage on a Jenkins agent. Both focused test suites pass successfully.

**Comment by MarkEWaite** [comment]
Thanks.  I don't understand why the changes to TarArchiver are needed, if the issue for the user is resolved by adjusting the exception in ArtifactArchiver.  Can you explain in more detail why the changes to TarArchiver are required?

**Comment by Hardik180704** [comment]
@MarkEWaite
 The TarArchiver change is needed because, for a non-empty unreadable file, the original AccessDeniedException does not reach ArtifactArchiver with the previous implementation.

TarArchiver previously wrote the entry header before opening the file. When opening the file failed, closeEntry() then failed because the declared number of bytes had not been written. That incomplete-entry exception masked the original AccessDeniedException, followed by the additional EOF/unclosed-entry errors seen in your reproduction.

As a result, changing only the AccessDeniedException handling in ArtifactArchiver works when that exception survives, such as the original zero-byte test case, but not for the non-empty file from your reproduction.

Opening the file before writing its tar header ensures that an unreadable file cannot create a partial entry. It also allows the original AccessDeniedException to reach ArtifactArchiver, where it can be converted into the actionable user-facing message.

**Comment by MarkEWaite** [comment]
I misdiagnosed a failure to encounter my breakpoints in the TarArchiver changes.  The TarArchiver class is run on the agent while my debugger was only attached to the Java process of the controller.  I inserted logging statements into the TarArchiver class at your changes and confirmed that they are called on the agent.

We'll need to run this change through the plugin compatibility tests and the acceptance test harness tests to assure that none of the larger test suites are broken by the change.

**Comment by Hardik180704** [comment]
@MarkEWaite
 Thanks for confirming that the TarArchiver changes are executed on the agent. Agreed that running the plugin compatibility and acceptance test harness suites is the appropriate next step. Please let me know if any failures require changes from my side.

File: core/src/test/java/hudson/util/io/TarArchiverTest.java [hunk 1]
```diff
@@ -131,6 +133,23 @@ void brokenSymlinks() throws Exception {
         }
     }
 
+    @Issue("JENKINS-27188")
```
Comment (MarkEWaite) [inline core/src/test/java/hudson/util/io/TarArchiverTest.java]: The issue reference format you used is for Jira, but this issue is only available from GitHub issues.
```suggestion
    @Issue("https://github.com/jenkinsci/jenkins/issues/27188")
```

File: test/src/test/java/hudson/tasks/ArtifactArchiverTest.java [hunk 2]
```diff
@@ -491,7 +492,7 @@ void fingerprintMigration() throws Exception {
     }
 
     @Test
-    @Issue("JENKINS-21905")
+    @Issue({"JENKINS-21905", "JENKINS-27188"})
```
Comment (MarkEWaite) [inline test/src/test/java/hudson/tasks/ArtifactArchiverTest.java]: The issue reference format you used is for Jira, but this issue is only available from GitHub issues.
```suggestion
    @Issue({"JENKINS-21905", "https://github.com/jenkinsci/jenkins/issues/27188"})
```

**Review by MarkEWaite — CHANGES_REQUESTED** [review]
2 changes for the issue references.

**Comment by Hardik180704** [comment]
@MarkEWaite 
Thanks for the correction. I updated both `@Issue` annotations to use the full GitHub URL for issue 27188 while retaining the Jira reference for JENKINS-21905.

I also reran `TarArchiverTest` and `ArtifactArchiverTest#archiveNotReadable`; both pass successfully.

**Review by MarkEWaite — APPROVED** [review]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback.  Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process.

/label ready-for-merge

**Comment by jdekeij** [comment]
Thank you very much for picking up the issue. Fix looks good and add useability to the product. 

I will report the changes to the build management so they can update the systems in the near future. 

Kind regards
Jasper de Keijzer
Rocket Software

**Comment by Hardik180704** [comment]
Thank you @MarkEWaite & @jdekeij 🥰
